### PR TITLE
Optimize CPVRRecording. No more copies, just shared pointers.

### PR DESCRIPTION
**Pull request by ksooo**
Followup to #5763. Same implementation approach...

Citing myself from #5763: "Next after this PR got merged will be optimization of CPVRRecording, which will by the way fix the issues with watched/unwatched state in recording window. @Jalle19 recently introduced a hack (#5613) that fixes immediate display of watched/unwatched in the recording window after changing the state via the respective context menu entries but that does not fix immediate display of watched state after a recording has been watched completely. The state will be set correctly, but it gets not displayed in the recording window, because the state is set at the wrong CPVRRecording instance. Eleminating all CPVRRecording copies will make Sam's hack obsolete (Recently introduced CPVRRecording::GetByFileItem which obtains the original PVRRecording belonging to a copy no longer needed than)."

### CODE REVIEW

File: xbmc/pvr/recordings/PVRRecording.cpp [hunk 1]
```diff
@@ -232,10 +232,7 @@ void CPVRRecording::UpdateMetadata(CVideoDatabase &db)
   if (!supportsPlayCount || !supportsLastPlayed)
   {
     if (!supportsPlayCount)
-    {
-       CFileItem pFileItem(*this);
-       m_playCount = db.GetPlayCount(pFileItem);
-    }
+      m_playCount = db.GetPlayCount(m_strFileNameAndPath);
```
Comment (xhaggi) [inline xbmc/pvr/recordings/PVRRecording.cpp]: indentation

Comment (xhaggi) [inline xbmc/pvr/recordings/PVRRecording.cpp]: ahh noo ;)

File: xbmc/video/VideoDatabase.cpp [hunk 2]
```diff
@@ -4627,10 +4627,9 @@ bool CVideoDatabase::GetPlayCounts(const std::string &strPath, CFileItemList &it
   return false;
 }
 
-int CVideoDatabase::GetPlayCount(const CFileItem &item)
+int CVideoDatabase::GetPlayCount(int iFileId)
```
Comment (xhaggi) [inline xbmc/video/VideoDatabase.cpp]: should we move this out in a separate commit?

**Comment by xhaggi** [comment]
nice work, looks good ;)

File: xbmc/video/VideoDatabase.cpp (hunk 2 above)
Comment (ksooo) [inline xbmc/video/VideoDatabase.cpp]: No. The diff is misleading. I did not replace GetPlayCount, but kept the existing 
https://github.com/xbmc/xbmc/pull/6207/files#diff-ee6ba71d6c3b070dfdba880abc06aeaeL409
and introduced two new
https://github.com/xbmc/xbmc/pull/6207/files#diff-ee6ba71d6c3b070dfdba880abc06aeaeR416 and https://github.com/xbmc/xbmc/pull/6207/files#diff-ee6ba71d6c3b070dfdba880abc06aeaeR880. The latter is private and now provides the actual implementation. The two others (reimplementation of existing method as well as the public new one) both use the private method. See here https://github.com/xbmc/xbmc/pull/6207/files#diff-dbf978b6cac348a2af93d3beec420e43R4659 and https://github.com/xbmc/xbmc/pull/6207/files#diff-dbf978b6cac348a2af93d3beec420e43R4664

**Comment by Jalle19** [comment]
jenkins build this please

**Comment by ksooo** [comment]
WIN-32 build error seams to be unrelated.

mv: cannot move `ffmpeg-2.5.2-Isengard-alpha' to `ffmpeg': Permission denied
./buildffmpeg.sh: line 35: cd: ffmpeg: No such file or directory
configuring ffmpeg
./buildffmpeg.sh: line 90: ./configure: No such file or directory
done
failed to compile /xbmc/system/players/dvdplayer/avcodec-56.dll
exiting msys environment
failed to build mingw libs
Build step 'Execute Windows batch command' marked build as failure

**Comment by Jalle19** [comment]
Yeah, jenkins has been acting up lately.

**Comment by Memphiz** [comment]
i am pretty sure jenkins is not acting up but there is a damned bug in one of those windows batches ...

**Comment by ksooo** [comment]
@opdenkamp could you please take a look and review/approve soon. I have more of this stuff in the queue... as you might know.

**Comment by opdenkamp** [comment]
this one will have to wait until tomorrow or the day after, took a look at the older ones first today :)

**Comment by ksooo** [comment]
@opdenkamp Ping?!

File: xbmc/FileItem.cpp [hunk 3]
```diff
@@ -382,19 +381,10 @@ const CFileItem& CFileItem::operator=(const CFileItem& item)
...
   }
 
-  if (item.HasPVRRecordingInfoTag())
-  {
-    m_pvrRecordingInfoTag = GetPVRRecordingInfoTag();
-    if (m_pvrRecordingInfoTag)
-      *m_pvrRecordingInfoTag = *item.m_pvrRecordingInfoTag;
-  }
+  if (item.m_pvrRecordingInfoTag)
+    m_pvrRecordingInfoTag = item.m_pvrRecordingInfoTag;
```
Comment (Jalle19) [inline xbmc/FileItem.cpp]: Could this not be simply `m_pvrRecordingInfoTag = item.m_pvrRecordingInfoTag` without the whole if-clause?

File: xbmc/filesystem/PVRFile.cpp [hunk 4]
```diff
@@ -75,7 +75,7 @@ bool CPVRFile::Open(const CURL& url)
     CFileItemPtr tag = g_PVRRecordings->GetByPath(strURL);
     if (tag && tag->HasPVRRecordingInfoTag())
     {
-      if (!g_PVRManager.OpenRecordedStream(*tag->GetPVRRecordingInfoTag()))
+      if (!g_PVRManager.OpenRecordedStream(tag->GetPVRRecordingInfoTag()))
```
Comment (Jalle19) [inline xbmc/filesystem/PVRFile.cpp]: Unrelated?

Comment (ksooo) [inline xbmc/filesystem/PVRFile.cpp]: No. https://github.com/xbmc/xbmc/pull/6207/files#diff-3fa68e30620ead600783e52c1987b7f4L322

Comment (Jalle19) [inline xbmc/filesystem/PVRFile.cpp]: Yeah, I realized later, I read it as `(*tag)->`

File: xbmc/pvr/addons/PVRClients.cpp [hunk 5]
```diff
@@ -1195,23 +1193,25 @@ bool CPVRClients::OpenStream(const CPVRChannel &tag, bool bIsSwitchingChannel)
...
-      client->OpenStream(tag))
+  if (tag)
   {
-    CSingleLock lock(m_critSection);
-    m_playingClientId = tag.m_iClientId;
-    m_bIsPlayingRecording = true;
-    m_strPlayingClientName = client->GetFriendlyName();
-    bReturn = true;
-  }
+    CloseStream();
```
Comment (Jalle19) [inline xbmc/pvr/addons/PVRClients.cpp]: This changes the logic, before `CloseStream` was called regardless of whether the tag was valid.

File: xbmc/FileItem.cpp (hunk 3 above)
Comment (ksooo) [inline xbmc/FileItem.cpp]: Not sure. Just kept the logic that was there before. Do assign recording info tag to this if the other item has a recording info tag, else do nothing. ?!

Comment (Jalle19) [inline xbmc/FileItem.cpp]: What I mean is now we basically have:

```
if(other.tag != NULL)
    tag = other.tag;
else
    tag = NULL;
```

Which can be shortened to one line.

**Comment by Jalle19** [comment]
@ksooo apart from my minor comments I think we're good to go. Better get things like this in sooner rather than later so we can catch any regressions.

File: xbmc/pvr/addons/PVRClients.cpp (hunk 5 above)
Comment (ksooo) [inline xbmc/pvr/addons/PVRClients.cpp]: Good catch. Will change that.

File: xbmc/FileItem.cpp (hunk 3 above)
Comment (ksooo) [inline xbmc/FileItem.cpp]: I think you are right, but the current implementation is buggy then, because it lets this' data untouched (even if it is not NULL) in case the other item has no recording info. other.recinfo = NULL => this.recinfo = (whatever-it-was-before, e.g. non-null). 

Do I have a brainfart or has the current implemenation really such a stupid bug?

Comment (Jalle19) [inline xbmc/FileItem.cpp]: Yeah, it seems if `item` (aka. "other) has a valid tag but we don't then we don't copy the tag over. That's definitely a bug, we should always copy the "other" tag, no matter what.

Comment (ksooo) [inline xbmc/FileItem.cpp]: Even worse, all the other *Tags are handled with the same wrong logic. I suggest to fix tem all togehter in a separate PR.

Comment (Jalle19) [inline xbmc/FileItem.cpp]: @ksooo I guess that would work, but at least make changes in separate commits (e.g. one for the channel tag and one for the timer).

**Comment by ksooo** [comment]
BTW: This should fix http://trac.kodi.tv/ticket/15229. Please add the "Fix" tag to this PR.

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by opdenkamp** [comment]
as said in another PR, haven't had a chance to review fully yet because 
the flu virus was demanding an audience.

**Comment by Jalle19** [comment]
@ksooo done

**Comment by Jalle19** [comment]
Can you squash to one commit?

**Comment by ksooo** [comment]
@jalle19 squashed down to a single commit, as requested.

File: xbmc/FileItem.cpp [hunk 6]
```diff
@@ -173,16 +173,18 @@ CFileItem::CFileItem(const CPVRChannel& channel)
...
   Initialize();
 
-  m_strPath = record.m_strFileNameAndPath;
   m_bIsFolder = false;
-  *GetPVRRecordingInfoTag() = record;
-  SetLabel(record.m_strTitle);
-  m_strLabel2 = record.m_strPlot;
-
+  m_pvrRecordingInfoTag = record;
+  if (record)
```
Comment (opdenkamp) [inline xbmc/FileItem.cpp]: you may assert() on this one imo, programming error to pass NULL in here

File: xbmc/FileItem.cpp [hunk 7]
```diff
@@ -490,8 +476,7 @@ void CFileItem::Reset()
   m_epgInfoTag.reset();
   delete m_pvrChannelInfoTag;
   m_pvrChannelInfoTag=NULL;
-  delete m_pvrRecordingInfoTag;
-  m_pvrRecordingInfoTag=NULL;
+  m_pvrRecordingInfoTag.reset();
```
Comment (opdenkamp) [inline xbmc/FileItem.cpp]: we better assign a new empty one here, in case one got copied, is used elsewhere, but is reset here.

File: xbmc/epg/EpgSearchFilter.cpp [hunk 8]
```diff
@@ -207,7 +207,7 @@ int EpgSearchFilter::FilterRecordings(CFileItemList &results)
   // TODO inefficient!
   for (int iRecordingPtr = 0; iRecordingPtr < recordings.Size(); iRecordingPtr++)
   {
-    CPVRRecording *recording = recordings.Get(iRecordingPtr)->GetPVRRecordingInfoTag();
+    const CPVRRecordingPtr recording = recordings.Get(iRecordingPtr)->GetPVRRecordingInfoTag();
```
Comment (opdenkamp) [inline xbmc/epg/EpgSearchFilter.cpp]: not 100% sure on this one, how the compiler treats this for shared_ptr, but aren't you just saying that the pointer is const here, instead of the referenced data? and if you want to make this a little bit more efficient, you move the declaration of this one 2 lines up, and just assign here ;-)

File: xbmc/addons/AddonCallbacksPVR.cpp [hunk 9]
```diff
@@ -191,8 +191,7 @@ void CAddonCallbacksPVR::PVRTransferRecordingEntry(void *addonData, const ADDON_
   }
 
   /* transfer this entry to the recordings container */
-  CPVRRecording transferRecording(*recording, client->GetID());
-  xbmcRecordings->UpdateFromClient(transferRecording);
+  xbmcRecordings->UpdateFromClient(*recording, client->GetID());
```
Comment (opdenkamp) [inline xbmc/addons/AddonCallbacksPVR.cpp]: no, add-on C types are not to be passed around. we translate to a CPVRRecording in here.

File: xbmc/interfaces/json-rpc/FileItemHandler.cpp [hunk 10]
```diff
@@ -359,7 +359,7 @@ void CFileItemHandler::HandleFileItem(const char *ID, bool allowFile, const char
     if (item->HasEPGInfoTag())
       FillDetails(item->GetEPGInfoTag().get(), item, fields, object, thumbLoader);
     if (item->HasPVRRecordingInfoTag())
-      FillDetails(item->GetPVRRecordingInfoTag(), item, fields, object, thumbLoader);
+      FillDetails(item->GetPVRRecordingInfoTag().get(), item, fields, object, thumbLoader);
```
Comment (opdenkamp) [inline xbmc/interfaces/json-rpc/FileItemHandler.cpp]: any reason for not changing FillDetails? or does it just take FileItem instances?

File: xbmc/pvr/addons/PVRClients.cpp [hunk 11]
```diff
@@ -1195,23 +1193,25 @@ bool CPVRClients::OpenStream(const CPVRChannel &tag, bool bIsSwitchingChannel)
...
+bool CPVRClients::OpenStream(const CPVRRecordingPtr &tag)
 {
   bool bReturn(false);
   CloseStream();
 
-  /* try to open the recording stream on the client */
-  PVR_CLIENT client;
-  if (GetConnectedClient(tag.m_iClientId, client) &&
-      client->OpenStream(tag))
+  if (tag)
```
Comment (opdenkamp) [inline xbmc/pvr/addons/PVRClients.cpp]: either return directly before CloseStream() or assert (cause programmer error), so this doesn't need another scope

File: xbmc/pvr/recordings/PVRRecording.h [hunk 12]
```diff
@@ -82,8 +84,23 @@ namespace PVR
     std::string   m_strFanartPath;    /*!< fanart path */
     unsigned      m_iRecordingId;     /*!< id that won't change while xbmc is running */
 
+  private:
```
Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: remove the private keyword for ctors here and the friend class, no reason to hack headers to prevent programmer errors. if you want to prevent others from creating instances, put it in a comment

File: xbmc/pvr/recordings/PVRRecording.h [hunk 13]
```diff
@@ -82,8 +84,23 @@ namespace PVR
...
     CPVRRecording(void);
     CPVRRecording(const PVR_RECORDING &recording, unsigned int iClientId);
+
+    // Prevent copy construction, even for CPVRRecording instances and friends.
+    // Note: Only declared, but intentionally not implemented
+    //       to prevent compiler generated copy ctor and to force
+    //       a linker error in case somebody tries to call it.
+    CPVRRecording(const CPVRRecording &tag);
```
Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: just implement them in the private part, then nobody can use it anyway. except when you use friend class hacks ;-)
every singleton in the code does it like this.

File: xbmc/pvr/recordings/PVRRecordings.cpp [hunk 14]
```diff
@@ -435,31 +435,26 @@ CPVRRecordingPtr CPVRRecordings::GetById(int iClientId, const std::string &strRe
...
-}
-
 void CPVRRecordings::Clear()
 {
   CSingleLock lock(m_critSection);
   m_recordings.clear();
 }
 
-void CPVRRecordings::UpdateFromClient(const CPVRRecording &tag)
+void CPVRRecordings::UpdateFromClient(const PVR_RECORDING &recording, unsigned int iClientId)
```
Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecordings.cpp]: this one needs to be changed back. c types belong in the add-on interface classes, not anywhere else in the code

File: xbmc/FileItem.cpp (hunk 7 above)
Comment (ksooo) [inline xbmc/FileItem.cpp]: Sorry, don't understand what you mean. What's the problem here?

BTW: Did it the same way for epginfotag and there it was kay for you. See some lines of code above...

File: xbmc/addons/AddonCallbacksPVR.cpp (hunk 9 above)
Comment (ksooo) [inline xbmc/addons/AddonCallbacksPVR.cpp]: Okay. Will change that.

File: xbmc/epg/EpgSearchFilter.cpp (hunk 8 above)
Comment (ksooo) [inline xbmc/epg/EpgSearchFilter.cpp]: Good catch. You're right regarding const. Will change this.

File: xbmc/FileItem.cpp (hunk 7 above)
Comment (opdenkamp) [inline xbmc/FileItem.cpp]: it got overlooked and needs to be changed too.

the problem is that we now no longer clear a copy of a tag, but clear the data in the reference to a tag.
fileitems are passed around all around the place, and Reset() can be called from about everything for whatever reason. so e.g. if we have some tag referenced in here, passed to a gui control which calls Reset() for whatever reason, we clear the tag in CPVRRecordings, instead of just the fileitem copy.

File: xbmc/interfaces/json-rpc/FileItemHandler.cpp (hunk 10 above)
Comment (ksooo) [inline xbmc/interfaces/json-rpc/FileItemHandler.cpp]: declaration: <code>void FillDetails(const ISerializable *info, ...</code>

BTW: Did it the same way as for epginfotag, were this was fine for you.

File: xbmc/pvr/addons/PVRClients.cpp (hunk 11 above)
Comment (ksooo) [inline xbmc/pvr/addons/PVRClients.cpp]: Okay. Will change this.

File: xbmc/interfaces/json-rpc/FileItemHandler.cpp (hunk 10 above)
Comment (opdenkamp) [inline xbmc/interfaces/json-rpc/FileItemHandler.cpp]: fine != overlooked, as with the other one :)
but sure, was just asking

File: xbmc/pvr/recordings/PVRRecording.h (hunk 12 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: Well, no. I'm trying to be very cooperative as you might have expirienced already, but doing it this way is good practice. 

Regarding private: May I friendly advice you to take a look Scott Meyers "More Effective C++" chapter 5.2. 
Regarding friend: From the architectural point of view CPVRRecordings is a factory for CPVRRecording instances. Thus, it is just fine to make it a friend.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 13 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: Scott Meyers again...

File: xbmc/pvr/recordings/PVRRecordings.cpp (hunk 14 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecordings.cpp]: Okay.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 12 above)
Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: Yeah I have read Meyers too while I was still studying, and I bet that this was relevant in 1995 when compilers didn't optimise this out, but I seriously doubt that it still is, and we use this throughout our code.

Even if it's more efficient after optimisations, we don't allocate a million of these, and not in a time critical thread. This just clutters.

File: xbmc/FileItem.cpp (hunk 7 above)
Comment (ksooo) [inline xbmc/FileItem.cpp]: So you're just questioning the sense of this PR as a whole? Wasn't it the intention no more to use copies and to always act upon the one end only instance? Confused... :-/

File: xbmc/interfaces/json-rpc/FileItemHandler.cpp (hunk 10 above)
Comment (ksooo) [inline xbmc/interfaces/json-rpc/FileItemHandler.cpp]: Will leave this as it is. Okay?

File: xbmc/FileItem.cpp (hunk 7 above)
Comment (opdenkamp) [inline xbmc/FileItem.cpp]: No I'm not, but not for everything that touches recordings. It makes perfect sense for methods that first created a FileItem copy, then passed it to GUI, which then had to find the correct recording in the list again based on the copy. But not so much for Reset(), which will very likely lead to issues, because nobody expects that calling Reset() on a fileitem clears out the contents of a recording in CPVRRecording...

File: xbmc/interfaces/json-rpc/FileItemHandler.cpp (hunk 10 above)
Comment (opdenkamp) [inline xbmc/interfaces/json-rpc/FileItemHandler.cpp]: yeah like I said, I was just asking why the .get() call was needed instead of passing the shared_ptr

File: xbmc/FileItem.cpp (hunk 7 above)
Comment (ksooo) [inline xbmc/FileItem.cpp]: Wait, I'm not quite sure whether I'm getting your point. shared_ptr::reset() does not touch the data of the object held by the shared_ptr instance, it (basically) just nulls its internal ptr member, right? So calling FileItem::Reset() which calls several .reset() then should not harm the objects held before the reset(). Or did I get some very basic stuff wrong?

Comment (opdenkamp) [inline xbmc/FileItem.cpp]: Ah, looks you're right and I just didn't look well enough :) Just looking at this one in a browser. As long as it doesn't touch the data, it's fine

File: xbmc/pvr/recordings/PVRRecording.h (hunk 12 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: Well ..... no. Don't get how this is related to compiler optimizations. This is about utilizing the compiler to help you preventing programming errors in the first place by restricting access privileges. And creating copies leads to all the problems and "find the original instance belonging to a copy" crap and many subtile bugs realated to the insane design the code now is suffering from. Don't see any drawbacks here, Please explain to me why this approach should be no good.

Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: Okay, I thought you meant that it would be more efficient.
You cannot create copies if the copy ctor and operator= are private and have a default implementation, like we do throughout the code, so I'm not sure what you're on about here. Just make the ctor public again so you don't need the friend class statement.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 13 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: BTW: These member functions are already declared in the private part. And only delcaring but not implementing them prevents not just friends from copying, but even instances of CPVRRecording itself. Maximum safety wrt. early error prevention, but no drawbacks here. Don't see what implementing these member functions has on the plus side.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 12 above)
Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: Perhaps my first comment and the one at https://github.com/xbmc/xbmc/pull/6207#discussion_r23719344 weren't clear enough now that I read it again. I didn't mean that you should make the copy ctor and operator= public, just the standard ctors, so we can do without the friend class statement. And there's no reason to leave the implementation of those out to enforce a linker error if the private keyword already prevents it from being copied, so you can remove the wall of text in the comment explaining why there's no implementation.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 13 above)
Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: That you can remove the wall of text, see comment above. And I suppose some linkers on some platform I don't use can bitch about this sort of thing too.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 12 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: Seams that we agree the having a private op= and copy ctor is fine. Good.

Regarding declaration only vs decl plus impl: Only declaring it has the advantage that friends and even instances of the class itself are prevented by the linker (yes, the linker, not the compiler) from creating copies. Thus, the point is not (only) to declare them private, but to (also)  NOT implement them at all. If you implement them, you have no chance to automatically detect accidentally copies generated by instances of the class itself (and friends ;-).

Regardig public ctor vs friend (seems to be your main concern): I also hate friend classes, but if the factory pattern can applied I'm fine with it. Sure, a friend always has access to everything in the class, but is giving everyone access to the ctor by making it public really better? I'm note quite sure about this...

File: xbmc/pvr/recordings/PVRRecording.h (hunk 13 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: lol re "wall of text". I thought that this kind of stuff needs explanation. In another project I know very well it would be enough to just write "// n.i." behind the copy ctor and op= declaration to indicate that this intentionally.

It IS the intention that linkers will bitch! This part of the pattern. If the method shall actually and accidentally called, what we want to prevent here, the linker will be unable to find an implementation for the method and exits with an error. Problem gets detected immediately, affected code can be easily detected and fixed.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 12 above)
Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: Question is, do we care that much about the class creating a copy instance of itself? I get what you're trying to do, but I can see no harm done if some code in CPVRRecording itself creates a copy really.

Re friend class: I consider them a hack, as they slash through all the protection that we put in place. Someone could touch private members that we don't expect to be changed from a friend class, and all those kinds of "nice" things. I can also see no problem if someone wants to allocate an instance of CPVRRecording. So we should avoid friend class statements if we can.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 13 above)
Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: as long as it only bitches on all platforms/linkers when actually being used, sure.

File: xbmc/pvr/recordings/PVRRecording.h (hunk 12 above)
Comment (davilla) [inline xbmc/pvr/recordings/PVRRecording.h]: that's why they had protected :) friends can't touch protected members (humm that sounds soooo dirty).

File: xbmc/pvr/recordings/PVRRecording.h (hunk 13 above)
Comment (Jalle19) [inline xbmc/pvr/recordings/PVRRecording.h]: For what it's worth I think that leaving these unimplemented is the right thing to do. Walls of texts are fine in headers, after all the header is supposed to describe what each function does. The whole idea here is to not accidentally create copies of objects we assume are one and the same and this seems like the perfect safeguard for that.

Comment (opdenkamp) [inline xbmc/pvr/recordings/PVRRecording.h]: see https://github.com/xbmc/xbmc/pull/6207/files#r23729933
as long as we don't get shit on some weird platform, sure. and it looks like @davilla said something too but deleted it again ;-)

Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: I red what @davilla wrote. It was just a short comment re "friend is evil", not "unimplemented cp-ctor and op=".

I'm pretty sure that all linkers on all relevant platforms will behave like expected, because the "Voodoo" stuff we're talking about is common practice since the ninetees. Looking up symbols and exiting with an error if not found is absolutely Basic linker functionality, nothing special at all.

I will keep this at it is now. If it bombs on any platform we will detect this early (e.g. Jenkins before merge) and I promise to implement those two member functions lightning fast. ;-)

File: xbmc/pvr/recordings/PVRRecording.h (hunk 12 above)
Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: Re friend class: ACK. I will come up with a solution without the friend class, most probably even without public ctor, because I'm still convinced that public ctor is not the right thing here and can be avoided.

Comment (ksooo) [inline xbmc/pvr/recordings/PVRRecording.h]: Okay. If a friend class is a nogo, solution will be to make ctors public again. Have no better idea. :-(

**Comment by ksooo** [comment]
All reviewer's feedback is now in:
~ use assert() instead of null checks at requested places
~ no more pass-around of add-on C data types
~ CPVRRecordings is no  longer a friend of class CPVRRecording :-)
~ normal ctors of class CPVRRecording are now public again (not cp-ctor and op=, which are private as before)
~text wallpapers in PVRRecording.h removed (okay, left a small comment there)

@opdenkamp: Thanks for the careful review. Hope, we're ready for merging now.

**Comment by ksooo** [comment]
@Captnoord +1. Possibly you should submit a PR addressing this consistency issue?

**Comment by Captnoord** [comment]
@ksooo +1 yea... me bitching about these kinda things makes you think.. 'why doesn't he do it himself?'
Maybe I will, I can't promise anything though.

**Comment by ksooo** [comment]
@opdenkamp is there anything I can do to speed things up? From my point of view I've done all I can do atm. ?!

**Comment by opdenkamp** [comment]
you could have patience, then at least you won't slow things down :)

**Comment by ksooo** [comment]
@opdenkamp hehe

**Comment by opdenkamp** [comment]
jenkins build and merge
